### PR TITLE
NPUW: Deref

### PR DESCRIPTION
**Pull request by dmatveev**
### Details:
 - ~~NPU plugin: Drop reference to `ov::Model` in the NPU plugin's `CompiledModel`~~ - 🚨🚨🚨- moved to a separate PR https://github.com/openvinotoolkit/openvino/pull/27777 
 - NPUW ext: Introduced a `detach_memory` step in the NPUW's `CompiledModel` to clear the original IR references if there's no fallback planned
 - NPUW ext: Fixed cyclic references in the online partitioning (preventing the original `ov::Model` from release)
 - NPUW ext: Fixed crashes in the updated `LazyTensor` implementation;
 - NPUW ext: Reworked `LazyTensor` a bit to have less code in & more predictable hashing behavior;
 - NPUW ext: Introduced `LazyTensor::detach()` & use it in the bank when LT is evaluated;
 - NPUW ext: Reworked the way how internal per-device memory banks are accessed - fixed a potential race.

### Tickets:
 - *E-142478*

### Related:
- https://github.com/openvinotoolkit/openvino.genai/pull/1259
- https://github.com/openvinotoolkit/openvino/pull/27777

### CODE REVIEW

**Review by dmatveev — COMMENTED** [review]
Self-review

File: src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp [hunk 1]
```diff
@@ -499,6 +500,25 @@ void ov::npuw::CompiledModel::finalize_weights_bank() {
...
+    for (size_t idx = 0; idx < m_compiled_submodels.size(); ++idx) {
+        auto& comp_model_desc = m_compiled_submodels[idx];
+        auto& proto_comp_model_desc = m_compiled_submodels[comp_model_desc.replaced_by.value_or(idx)];
+        if (!proto_comp_model_desc.model || !proto_comp_model_desc.compiled_model) {
+            continue; // optimized-out OR already cleared - skip
+        }
+        if (proto_comp_model_desc.device_it + 1 == m_dev_list.end()) {
+            LOG_INFO("No fallback expected - clear the OV model for Subgraph[" << idx << "]");
+            proto_comp_model_desc.model.reset();
+            // proto_comp_model_desc.compiled_model = {}; // Shouldn't be here, CPU only
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp]: UPD. Also the case for CPU

File: src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp [hunk 2]
```diff
@@ -499,6 +500,23 @@ void ov::npuw::CompiledModel::finalize_weights_bank() {
     LOG_INFO("Done.");
 }
 
+void ov::npuw::CompiledModel::detach_memory() {
+    LOG_INFO("Detaching model & weight memory...");
+    LOG_BLOCK();
+    for (size_t idx = 0; idx < m_compiled_submodels.size(); ++idx) {
+        auto& comp_model_desc = m_compiled_submodels[idx];
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp:507]: Maybe, as an alternative to calling `.detach()` in the bank itself, it could've been called here for lazy tensors in the closure - @smirnov-alexey consider this change for the future

Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp:507]: I had it initially but it required additional checks. I prefer this implementation since it's desired behavior for every case anyway

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp [hunk 3]
```diff
@@ -200,17 +214,25 @@ ov::Tensor LazyTensorImpl::eval() const {
...
+
+std::size_t LazyTensorImpl::get_hash() const {
+    return m_hash;
+}
 
-    ov::Tensor result = std::visit(overloaded{[](const auto& op) {
-                                       return op.eval();
-                                   }},
-                                   m_transform);
-    NPUW_ASSERT(result);
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:208]: I'd keep this assert but fine either way

File: src/plugins/intel_npu/src/plugin/npuw/util.hpp [hunk 4]
```diff
@@ -141,6 +141,14 @@ Impl<M> _(std::shared_ptr<M> pM) {
 
 }  // namespace at
 
+// Written here to be a drop-in replacement for ov::parallel_for for the debug purposes
+template <typename F>
+void non_parallel_for(std::size_t count, F&& f) {
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/util.hpp:146]: Are we keeping it?

**Review by smirnov-alexey — APPROVED** [review]

File: src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp (hunk 2 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp:507]: However in case of having some CPU-compiled subgraphs I'd implement something like this https://github.com/openvinotoolkit/openvino/pull/27415/files#diff-2ed7db3ac71f25124dadd21bb7a90ec92b2ba192cbf9a864e83828ba77ef6926R565 - we need to copy CPU consts to and reset the ones from the original model

Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp:507]: Can be a FIXME for the future

File: src/plugins/intel_npu/src/plugin/npuw/util.hpp (hunk 4 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/util.hpp:146]: Yes, moved here to stop reinventing it all the time
